File: nostr-knowledge/examples/basics/verify-signature/verify_signature.py
"""
NOSTR Verify Event Signature — Python

Takes a NOSTR event (as JSON), recomputes the event ID from the canonical
serialization, and verifies the Schnorr signature (BIP-340).

Dependencies:
    pip install pynostr

Run:
    python verify_signature.py

References:
    - NIP-01: https://github.com/nostr-protocol/nips/blob/master/01.md
"""

import hashlib
import json
import time

from pynostr.key import PrivateKey
from pynostr.event import Event

print("=== NOSTR Verify Event Signature ===\n")

# --- Create a sample event to verify ---
# In production, you'd receive this from a relay or another client.
private_key = PrivateKey()
public_key = private_key.public_key

event = Event(
    pubkey=public_key.hex(),
    created_at=int(time.time()),
    kind=1,
    tags=[],
    content="This is a signed NOSTR event for verification testing.",
)
private_key.sign_event(event)

print("Sample event to verify:")
print(json.dumps(event.to_dict(), indent=2))
print()

# --- Step 1: Manually recompute the event ID ---
# The ID is SHA-256 of the canonical serialization:
#   [0, <pubkey>, <created_at>, <kind>, <tags>, <content>]
serialization = json.dumps(
    [0, event.pubkey, event.created_at, event.kind, event.tags, event.content],
    separators=(",", ":"),
    ensure_ascii=False,
)
computed_id = hashlib.sha256(serialization.encode("utf-8")).hexdigest()

print("--- Step 1: Verify Event ID ---")
print(f"  Serialized: {serialization}")
print(f"  Computed ID:  {computed_id}")
print(f"  Event ID:     {event.id}")
print(f"  ID matches:   {computed_id == event.id}")
print()

# --- Step 2: Verify the Schnorr signature ---
# The event's verify() method checks the BIP-340 Schnorr signature.
# It verifies that event.sig is a valid signature over event.id
# using the public key event.pubkey.
is_valid = event.verify()

print("--- Step 2: Verify Schnorr Signature ---")
print(f"  Signature valid: {is_valid}")
print()

# --- Step 3: Demonstrate what happens with a tampered event ---
print("--- Tampering Test ---")

# Create a tampered event with modified content but original ID and sig.
tampered = Event(
    pubkey=event.pubkey,
    created_at=event.created_at,
    kind=event.kind,
    tags=event.tags,
    content="TAMPERED CONTENT",
)
# Manually set the original ID and signature (simulating tampering).
tampered.id = event.id
tampered.sig = event.sig

# Recompute the ID for the tampered event.
tampered_serialization = json.dumps(
    [0, tampered.pubkey, tampered.created_at, tampered.kind, tampered.tags, tampered.content],
    separators=(",", ":"),
    ensure_ascii=False,
)
tampered_computed_id = hashlib.sha256(tampered_serialization.encode("utf-8")).hexdigest()

print(f'  Original content: "{event.content}"')
print(f'  Tampered content: "{tampered.content}"')
print(f"  Tampered computed ID: {tampered_computed_id}")
print(f"  Stored ID:            {tampered.id}")
print(f"  ID matches: {tampered_computed_id == tampered.id}")
print()

# --- Step 4: Explain the full verification flow ---
print("--- Verification Steps ---")
print("1. Serialize: [0, pubkey, created_at, kind, tags, content]")
print("2. Compute SHA-256 of the serialized JSON string")
print("3. Compare computed hash with event.id")
print("4. Verify Schnorr signature (event.sig) over event.id using event.pubkey")
print("5. If ID matches AND signature is valid -> event is authentic")

File: nostr-knowledge/examples/basics/verify-signature/verify_signature.go
/*
NOSTR Verify Event Signature — Go

Takes a NOSTR event, recomputes the event ID from the canonical serialization,
and verifies the Schnorr signature (BIP-340).

Dependencies:
	go get github.com/nbd-wtf/go-nostr

Run:
	go run verify_signature.go

References:
	- NIP-01: https://github.com/nostr-protocol/nips/blob/master/01.md
*/

package main

import (
	"encoding/json"
	"fmt"
	"log"
	"time"

	"github.com/nbd-wtf/go-nostr"
)

func main() {
	fmt.Println("=== NOSTR Verify Event Signature ===")
	fmt.Println()

	// --- Create a sample event to verify ---
	// In production, you'd receive this from a relay or another client.
	secretKey := nostr.GeneratePrivateKey()
	publicKey, err := nostr.GetPublicKey(secretKey)
	if err != nil {
		log.Fatalf("Failed to derive public key: %v", err)
	}

	event := nostr.Event{
		PubKey:    publicKey,
		CreatedAt: nostr.Timestamp(time.Now().Unix()),
		Kind:      1,
		Tags:      nostr.Tags{},
		Content:   "This is a signed NOSTR event for verification testing.",
	}

	err = event.Sign(secretKey)
	if err != nil {
		log.Fatalf("Failed to sign event: %v", err)
	}

	eventJSON, _ := json.MarshalIndent(event, "", "  ")
	fmt.Println("Sample event to verify:")
	fmt.Println(string(eventJSON))
	fmt.Println()

	// --- Step 1: Verify the event ID ---
	// go-nostr's GetID() recomputes the ID from the canonical serialization:
	//   SHA-256([0, pubkey, created_at, kind, tags, content])
	computedID := event.GetID()

	fmt.Println("--- Step 1: Verify Event ID ---")
	fmt.Printf("  Computed ID: %s\n", computedID)
	fmt.Printf("  Event ID:    %s\n", event.ID)
	fmt.Printf("  ID matches:  %v\n\n", computedID == event.ID)

	// --- Step 2: Verify the Schnorr signature ---
	// CheckSignature() verifies the BIP-340 Schnorr signature.
	// It checks that event.Sig is valid over event.ID using event.PubKey.
	valid, err := event.CheckSignature()
	if err != nil {
		log.Fatalf("Signature check error: %v", err)
	}

	fmt.Println("--- Step 2: Verify Schnorr Signature ---")
	fmt.Printf("  Signature valid: %v\n\n", valid)

	// --- Step 3: Demonstrate what happens with a tampered event ---
	fmt.Println("--- Tampering Test ---")

	// Create a copy with modified content but original ID and sig.
	tampered := event
	tampered.Content = "TAMPERED CONTENT"

	// The ID is no longer correct because content changed.
	tamperedComputedID := tampered.GetID()

	fmt.Printf("  Original content: \"%s\"\n", event.Content)
	fmt.Printf("  Tampered content: \"%s\"\n", tampered.Content)
	fmt.Printf("  Tampered computed ID: %s\n", tamperedComputedID)
	fmt.Printf("  Stored ID:           %s\n", tampered.ID)
	fmt.Printf("  ID matches: %v\n\n", tamperedComputedID == tampered.ID)

	// Even if we try to verify the signature, it will fail because
	// the ID doesn't match, and the signature was over the original ID.
	tamperedValid, _ := tampered.CheckSignature()
	fmt.Printf("  Tampered signature check: %v (expected false)\n\n", tamperedValid)

	// --- Step 4: Explain the verification flow ---
	fmt.Println("--- Verification Steps ---")
	fmt.Println("1. Serialize: [0, pubkey, created_at, kind, tags, content]")
	fmt.Println("2. Compute SHA-256 of the serialized JSON string")
	fmt.Println("3. Compare computed hash with event.id")
	fmt.Println("4. Verify Schnorr signature (event.sig) over event.id using event.pubkey")
	fmt.Println("5. If ID matches AND signature is valid -> event is authentic")
}
